File: QUICKSORT/quicksort.c
/* Implementación de ALgoritmo Quicksort
    Desarrollado por: Casiano Granados Brandon Antonio */
#include <stdlib.h>
#include <stdio.h>
#define MAX 20
 
 void intercambio(int [],int, int);
 void quicksort(int [],int, int); 
 void imprimeVector(int [],int);
 
 int main()
 {
    int vector[MAX]={0};
    int numElem,i;
    clrscr();
    printf("\nCuantos elementos quieres en el arreglo? ");
    scanf("%d",&numElem);
    for(i=0;i<numElem;i++)
    { 
        printf("\nDato: ");
        scanf("%d",&vector[i]);
    }
 
    quicksort(vector,0,numElem-1);
    imprimeVector(vector,numElem);
    getch();
    return 0;
 }
 
 //Función que se encarga de cambiar los valores
 void intercambio(int vector[],int i, int j)
 {
     int aux;
    aux=vector[i];
    vector[i]=vector[j];
    vector[j]=aux;
    return;
 }
//Función que se encarga de ordenar los valores con el Algoritmo de
//manera iterativa
 void quicksort(int vector[],int inf, int sup)
 {
    int i=inf, j=sup, x;
    x=vector[(i+j)/2]; //obteniendo el pivote
    while(i<=j)
    { 
        while(vector[i]<x)    //obteniendo elemento mayor de lado izquierdo del  pivote para intercambiar
            i++;
        while(vector[j]>x)   // obteniendo elemento menor de lado derecho del pivote a intercambiar
            j--;
        if(i<=j)             //intercambiando cuando se sigan presentando valores menores ala derecha y mayores ala izquierda
        {
            intercambio(vector,i,j);
            i++;
            j--;
         }
    }
    //Llamamos a la función quicksort para ordenar los elementos menores al pivote
    if(inf<j)
        quicksort(vector,inf,j);
    //Llamamos a la función quicksort para ordenar los elementos mayores al pivote
    if(i<sup)
        quicksort(vector,i,sup);
        return;
 }
 /* Función que se encarga de imprimir el los arreglos */
 void imprimeVector(int vector[],int elem)
 {
    int i;
    printf("\nVector ordenado con quicksort\n");
    for(i=0;i<elem;i++)
        printf("[%d] ",vector[i]);
    return;
 }

File: QUICKSORT-JAVA/quicksort.java
/* Implementación de Algoritmo Quicksort
 * Desarrollado por: Casiano Granados Brandon Antonio
  */
public class quicksort{
    public static void main (String args[])
    {
        int arreglo[] = {10,8,5,7,1,3,6,2,9,4};
        int tam = arreglo.length-1;
        imprime(arreglo);
        qs(arreglo, 0, tam);
        imprime(arreglo);
        return;
    }
    /* Función encargada de imprimir los arreglos */
    private static void imprime(int imp[])
    {
        for(int i=0; i<imp.length; i++)
            System.out.println(imp[i]);
        return;
    }
    /* Función que ordena los arreglos con el algoritmo */
    private static void qs(int vector[], int inf, int sup)
    {
        int izq, der, pivote, aux;
        izq = inf;
        der = sup;
        pivote = vector[(izq + der)/2];
        do
        {
            while((vector[izq] < pivote) && (izq < sup))//Aquí se revisa si el vector es menor al pivote
                izq++;
            while((vector[der] > pivote) && (der > inf))//Aquí se revisa si el vector es maqyor al pivote
                der--;
            if(izq <= der)//Aquí cambiamos los valores para ordenar
            {
                aux = vector[izq];
                vector[izq] = vector[der];
                vector[der] = aux;
                izq++;
                der++;
            }
        }while(izq <= der);
        //Con los arreglos de valores menores y mayores se envían a la misma función para ordenar
        if(inf < der)
            qs(vector, inf, der);
        if(izq < sup)
            qs(vector, izq, sup);
        return;
    }
}